### PR TITLE
feat: add transform / filter / write consumer lambda

### PR DESCRIPTION
**Pull request by olizilla**
Queue consumer lambda to transform items to the current index format, check if they exists in destination table, and write those records that are missing.

Each record is a stringified array of up to 500 BlocksIndex objects from the Scanner.

Any failed writes from the DynamoDB BatchWriteCommand are send to the `unprocessedWritesQueue` for debugging and re-driving.

Any other errors and the message is released back to the queue. If that batch fails 3 times, it is send to the `batchDeadLetterQueue`

License: MIT

### CODE REVIEW

**Comment by olizilla** [comment]
- The `unprocessedWritesQueue` will contain stringified arrays of `WriteRequest` objects, the marshalled Dynamo DB BulkWriteCommand format
- The `batchDeadLetterQueue` will contain stringified arrays of `BlockIndex` objects, the same format as the `batchQueue`, so anything in the could be re-driven through the consumer lambda if there is no obvious, intrinsic reason why that batch failed

File: packages/functions/src/consumer.ts [hunk 1]
```diff
@@ -1,15 +1,175 @@
...
+
+  const sqs = new SQSClient({})
+  const dynamo = new DynamoDBClient({})
+
+  const records: any[] = event.Records
   for (const r of records) {
-    const batch = JSON.parse(r.body)
-    console.log(`Batch of ${batch.length}`)
+    const items: BlocksIndex[] = JSON.parse(r.body)
+    console.log(`Processing "${items.length}" items`)
```
Comment (alanshaw) [inline packages/functions/src/consumer.ts]: lol quotes

File: packages/functions/src/consumer.ts [hunk 2]
```diff
@@ -1,15 +1,175 @@
...
+ * Ideally we'd use a conditional write, but amazon charges the write cost 
+ * regardless, even if the condition prevents the write.
+ * 
+ * We have to manually avoid duplicate multihash + carpath keys per batch:
+ * > ValidationException: Provided list of item keys contains duplicates
+ * 
+ * We ignore any UnprocessedItems from the BatchGetItem response. It will
+ * be interpreted as having not existed, and so cause a possibly un-needed write.
+ */
+export async function* checkExists(dstTable: string, client: DynamoDBClient, items: BlocksCarsPosition[]) {
```
Comment (alanshaw) [inline packages/functions/src/consumer.ts]:
```suggestion
export async function* filterExists(dstTable: string, client: DynamoDBClient, items: BlocksCarsPosition[]) {
```

**Review by alanshaw — APPROVED** [review]
Looks great!

**Comment by olizilla** [comment]
Running with the staging blocks table as the src and a new empty table as the dst, i get 2 interesting issues

1. The dst table complains "ProvisionedThroughputExceededException: The level of configured provisioned throughput for the table was exceeded. Consider increasing your provisioning level with the UpdateTable API."
2. The consumer lambda errors but the messages it is operating on are not marked as visible again. My guess is I am waiting for the 15 visibility timeout until those messages are retried.

Screenshot 2023-09-07 at 17 03 44

**Pull request merged**